File: l2-neet150/trees/1448.java
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public int goodNodes(TreeNode root) {
        return goodNodes(root, Integer.MIN_VALUE); // !! my fault, not 0, cause vals doesn't start from 0
    }

    public int goodNodes(TreeNode root, int prevMax) {
        if (root == null) {
            return 0;
        }

        boolean isGood = (root.val >= prevMax);
        int max = Math.max(prevMax, root.val);
        int countLeft = goodNodes(root.left, max);
        int countRight = goodNodes(root.right, max);
        int total = countLeft + countRight + (isGood ? 1 : 0);

        return total;
    }
}

File: _py/1448.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        def goodNodes(self, root: TreeNode, maxSoFar: int) -> int:
            if root is None:
                return 0
            count = 0
            if maxSoFar <= root.val:
                count += 1
                maxSoFar = root.val
            count += goodNodes(self, root.left, maxSoFar)
            count += goodNodes(self, root.right, maxSoFar)
            return count

        return goodNodes(self, root, root.val)
